File: python/src/coasty/sse.py
"""Server-Sent Events: parser + reconnecting iterator with Last-Event-ID.

Framing (see API_NOTES "SSE framing reference"):

- UTF-8 lines, events separated by a blank line.
- Fields: ``id: <seq>``, ``event: <type>``, ``data: <json>``; multiple
  ``data:`` lines are joined with ``\\n``.
- Lines starting with ``:`` are keepalive comments and are ignored.
- The client tracks the last seen ``id`` and sends it as ``Last-Event-ID``
  on reconnect; events are durable server-side so the seq cursor guarantees
  no loss. Duplicates (replays at/below the cursor) are dropped client-side.
"""

from __future__ import annotations

import json
import time
from collections.abc import Callable, Iterable, Iterator
from dataclasses import dataclass
from typing import Any

import httpx

from .errors import CoastyError

DONE_EVENT = "done"


class StreamInterruptedError(CoastyError):
    """The event stream kept dropping before a terminal ``done`` event."""

    default_code = "STREAM_INTERRUPTED"
    default_type = "transport_error"


@dataclass(frozen=True)
class SSEEvent:
    """One parsed SSE event (``seq`` is the numeric form of ``id``)."""

    event: str
    data: str
    id: str | None = None

    @property
    def seq(self) -> int | None:
        if self.id is not None and self.id.isdigit():
            return int(self.id)
        return None

    def json(self) -> Any:
        """Decode the ``data`` payload as JSON."""
        return json.loads(self.data)


def parse_sse_lines(lines: Iterable[str]) -> Iterator[SSEEvent]:
    """Parse decoded SSE lines (no trailing newlines) into events.

    Multi-line ``data:`` is joined with newlines, ``:`` comments are ignored,
    and a partial event at EOF (no terminating blank line) is discarded, per
    the SSE spec -- the reconnect replay will redeliver it.
    """
    event_id: str | None = None
    event_type: str | None = None
    data_lines: list[str] = []
    pending = False

    for raw in lines:
        line = raw.rstrip("\r\n")
        if line == "":
            if pending:
                yield SSEEvent(
                    event=event_type or "message",
                    data="\n".join(data_lines),
                    id=event_id,
                )
            event_type = None
            data_lines = []
            pending = False
            continue
        if line.startswith(":"):
            continue  # comment / keepalive
        field, _, value = line.partition(":")
        if value.startswith(" "):
            value = value[1:]
        if field == "id":
            event_id = value
            pending = True
        elif field == "event":
            event_type = value
            pending = True
        elif field == "data":
            data_lines.append(value)
            pending = True
        # unknown fields (incl. "retry") are ignored


def iter_events_reconnecting(
    open_stream: Callable[[str | None], Iterator[str]],
    *,
    last_event_id: str | None = None,
    done_event: str = DONE_EVENT,
    max_reconnects: int = 5,
    reconnect_delay: float = 0.5,
    sleep: Callable[[float], None] = time.sleep,
) -> Iterator[SSEEvent]:
    """Yield events, transparently reconnecting via ``Last-Event-ID``.

    ``open_stream(cursor)`` must open a fresh stream of decoded lines,
    sending ``cursor`` as the ``Last-Event-ID`` header when not ``None``.
    The iterator stops cleanly after ``done_event``; a stream that ends or
    drops earlier is reopened (up to ``max_reconnects`` times) and replayed
    events at or below the seq cursor are skipped, so callers see every
    event exactly once.
    """
    cursor = last_event_id
    last_seq = int(cursor) if cursor is not None and cursor.isdigit() else None
    reconnects = 0

    while True:
        try:
            for event in parse_sse_lines(open_stream(cursor)):
                if event.id is not None:
                    cursor = event.id
                seq = event.seq
                if seq is not None:
                    if last_seq is not None and seq <= last_seq:
                        continue  # duplicate from a replay
                    last_seq = seq
                yield event
                if event.event == done_event:
                    return
        except (httpx.TransportError, ConnectionError, TimeoutError):
            pass  # dropped mid-stream: fall through to reconnect

        reconnects += 1
        if reconnects > max_reconnects:
            raise StreamInterruptedError(
                f"event stream ended before a '{done_event}' event "
                f"after {max_reconnects} reconnect(s)",
            )
        if reconnect_delay > 0:
            sleep(reconnect_delay)

File: typescript/src/coasty/sse.ts
/**
 * Server-Sent Events parsing + reconnection.
 *
 * Framing (UTF-8): events are separated by a blank line; `id:`/`event:`/`data:`
 * lines; multiple `data:` lines join with `\n`; lines starting with `:` are
 * comments (keepalives) and are ignored. The client tracks the last seen `id`
 * and sends it as `Last-Event-ID` on reconnect — Coasty replays everything
 * after that seq with no loss and no duplicates.
 */

export interface SseMessage {
  /** Last seen `id:` value (the durable seq for Coasty streams). */
  id: string | null;
  /** `event:` field, defaulting to "message" per the SSE spec. */
  event: string;
  /** Joined `data:` payload. */
  data: string;
}

export type SleepFn = (ms: number) => Promise<void>;

export const defaultSleep: SleepFn = (ms) =>
  new Promise((resolve) => {
    setTimeout(resolve, ms);
  });

/**
 * Parse a raw SSE byte stream into messages. Incomplete trailing events (no
 * blank-line terminator before EOF) are discarded, per the SSE spec.
 */
export async function* parseSseStream(
  stream: ReadableStream<Uint8Array>,
): AsyncGenerator<SseMessage, void, undefined> {
  const reader = stream.getReader();
  const decoder = new TextDecoder('utf-8');

  let buffer = '';
  let dataLines: string[] = [];
  let eventType = '';
  let lastId: string | null = null;
  let sawField = false;

  const flush = (): SseMessage | null => {
    if (!sawField || dataLines.length === 0) {
      // Per spec: a blank line with an empty data buffer dispatches nothing.
      dataLines = [];
      eventType = '';
      sawField = false;
      return null;
    }
    const message: SseMessage = {
      id: lastId,
      event: eventType === '' ? 'message' : eventType,
      data: dataLines.join('\n'),
    };
    dataLines = [];
    eventType = '';
    sawField = false;
    return message;
  };

  const handleLine = (line: string): SseMessage | null => {
    if (line === '') return flush();
    if (line.startsWith(':')) return null; // comment / keepalive
    const colon = line.indexOf(':');
    const field = colon === -1 ? line : line.slice(0, colon);
    let value = colon === -1 ? '' : line.slice(colon + 1);
    if (value.startsWith(' ')) value = value.slice(1);
    switch (field) {
      case 'id':
        // Per spec, ids containing NUL are ignored.
        if (!value.includes('\0')) {
          lastId = value;
          sawField = true;
        }
        break;
      case 'event':
        eventType = value;
        sawField = true;
        break;
      case 'data':
        dataLines.push(value);
        sawField = true;
        break;
      default:
        // Unknown fields (e.g. "retry") are ignored.
        break;
    }
    return null;
  };

  try {
    for (;;) {
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });
      // Normalize CRLF / CR line endings, then emit complete lines.
      for (;;) {
        const match = /\r\n|\n|\r/.exec(buffer);
        if (match === null) break;
        const line = buffer.slice(0, match.index);
        buffer = buffer.slice(match.index + match[0].length);
        const message = handleLine(line);
        if (message !== null) yield message;
      }
    }
    // EOF: a final line without a trailing newline could close a frame only
    // via a blank line, which requires a newline — so leftover buffer content
    // belongs to an unterminated (discarded) event.
  } finally {
    try {
      await reader.cancel();
    } catch {
      // The stream may already be closed/errored; cancellation is best-effort.
    }
  }
}

/** Returns a fresh SSE byte stream, resuming after `lastEventId` when given. */
export type SseConnector = (lastEventId: string | null) => Promise<ReadableStream<Uint8Array>>;

export interface ReconnectingSseOptions {
  /** Max reconnect attempts after a dropped stream (default 5). */
  maxReconnects?: number;
  /** Delay between reconnect attempts in ms (default 500). */
  reconnectDelayMs?: number;
  sleep?: SleepFn;
  /** Stream terminator (default: the documented `done` event). */
  isTerminal?: (message: SseMessage) => boolean;
  /** Resume cursor carried into the first connection. */
  lastEventId?: string | null;
}

/** Raised when the stream keeps dropping without ever reaching `done`. */
export class SseStreamEndedError extends Error {
  constructor(reconnects: number) {
    super(
      `SSE stream ended without a 'done' event after ${String(reconnects)} reconnect attempt(s)`,
    );
    this.name = 'SseStreamEndedError';
  }
}

/**
 * Async iterator over SSE messages that transparently reconnects when the
 * stream drops mid-flight, resuming via `Last-Event-ID`. Numeric ids are used
 * as a monotonic cursor: replayed events with `seq <= last seen` are filtered
 * out, so consumers observe no loss and no duplicates. Terminates after the
 * `done` event.
 */
export async function* reconnectingSse(
  connect: SseConnector,
  options: ReconnectingSseOptions = {},
): AsyncGenerator<SseMessage, void, undefined> {
  const maxReconnects = options.maxReconnects ?? 5;
  const reconnectDelayMs = options.reconnectDelayMs ?? 500;
  const sleep = options.sleep ?? defaultSleep;
  const isTerminal = options.isTerminal ?? ((message: SseMessage) => message.event === 'done');

  let lastEventId: string | null = options.lastEventId ?? null;
  let lastSeq: number | null =
    lastEventId !== null && /^\d+$/.test(lastEventId) ? Number(lastEventId) : null;
  let reconnects = 0;

  for (;;) {
    // Connection errors (e.g. 401/404 envelopes) propagate to the caller.
    const stream = await connect(lastEventId);
    let sawTerminal = false;
    try {
      for await (const message of parseSseStream(stream)) {
        if (message.id !== null) {
          if (/^\d+$/.test(message.id)) {
            const seq = Number(message.id);
            if (lastSeq !== null && seq <= lastSeq) continue; // replayed duplicate
            lastSeq = seq;
          }
          lastEventId = message.id;
        }
        yield message;
        if (isTerminal(message)) {
          sawTerminal = true;
          break;
        }
      }
    } catch {
      // Mid-stream transport failure: treat as a drop and reconnect below.
    }
    if (sawTerminal) return;
    reconnects += 1;
    if (reconnects > maxReconnects) throw new SseStreamEndedError(maxReconnects);
    await sleep(reconnectDelayMs);
  }
}
